### PR TITLE
support DataLoader with multi-process mode on MacOs and Windows basically

### PR DESCRIPTION
**Pull request by wwqgtxx**
### PR types
New features

### PR changes

APIs

### Describe

After reading code, I found we can simply comment out some platform-related code to make DataLoader with multi-process mode work on MacOs and Windows.

### CODE REVIEW

**Comment by luotao1** [comment]
Sorry for too late to notice this PR, could you resolve the conflict files to pass the CI first? @wwqgtxx

**Comment by heavengate** [comment]
DataLoader unittest is disabled for Mac/Windows currently, could you please remove these lines to reopen DataLoader unites for Mac/Windows to run unitests CI
https://github.com/PaddlePaddle/Paddle/blob/b93b1e34aa1b45ecc64bad259412078e8904d44d/python/paddle/fluid/tests/unittests/CMakeLists.txt#L480

**Comment by wwqgtxx** [comment]
maybe `test_multiprocess_dataloader_dataset (Timeout)` cause by #44170 #36819

**Comment by luotao1** [comment]
Do you try `test_multiprocess_dataloader_dataset` successfully on your local mac or windows? You can adjust the elapsed time such as `set_tests_properties(jit_kernel_test PROPERTIES TIMEOUT 120)` to pass the CI.

**Comment by wwqgtxx** [comment]
> Do you try `test_multiprocess_dataloader_dataset` successfully on your local mac or windows? You can adjust the elapsed time such as `set_tests_properties(jit_kernel_test PROPERTIES TIMEOUT 120)` to pass the CI.

In fact, when trying to use `multiprocessing` to pass `Tensor`, there must be an error. 
I ran the test in `test_multiprocess_dataloader_dataset` locally, and found that the test of passing pure numpy arrays can pass normally. Once a `Tensor` appears in the return value of the `DataSet.__iter__`, it will fail. 
This behavior is the same as #44170. So this should not be a problem caused by this PR.

**Comment by luotao1** [comment]
@wwqgtxx Discussed with @heavengate:
1. You can remove `list(REMOVE_ITEM TEST_OPS test_multiprocess_dataloader_dataset)` to pass the CI
2. After all CI passed, we will merge this PR
3. Maybe you could help us to fix #44170

File: python/paddle/fluid/reader.py [hunk 1]
```diff
@@ -481,17 +481,17 @@ def __init__(self,
...
         self.num_workers = num_workers
 
         assert prefetch_factor > 0, "prefetch_factor should be a positive value"
 
         self.use_shared_memory = use_shared_memory
+        if use_shared_memory and (sys.platform == 'darwin'
+                                  or sys.platform == 'win32'):
+            warnings.warn(
+                "DataLoader with multi-process mode is not fully supported on MacOs and Windows currently." \
+                " Please use multi-process mode with use_shared_memory = False instead")
```
Comment (luotao1) [inline python/paddle/fluid/reader.py:493]: @wwqgtxx 
现在CI还有多个单测失败，可以继续修一下么？比如这段就会触发`test_dataloader_autotune`失败
https://xly.bce.baidu.com/paddlepaddle/paddle/newipipe/detail/6548861/job/18199166
```
2022-09-06 16:54:46 C:\home\workspace\Paddle\build\python\paddle\fluid\reader.py:492: UserWarning: DataLoader with multi-process mode is not fully supported on MacOs and Windows currently. Please use multi-process mode with use_shared_memory = False instead
2022-09-06 16:54:46   "DataLoader with multi-process mode is not fully supported on MacOs and Windows currently." \
2022-09-06 16:54:46 test_dataloader_autotune failed
```

Comment (wwqgtxx) [reply python/paddle/fluid/reader.py:493]: 其实我也看过ci的日志，大部分单测失败都是因为某个class不能被pickle，就像之前提到的Tensor一样，这个问题可能是因为paddle的一些c/c++代码实现的基础类型就是不能被pickle导致的。
个人感觉作为一个`basically`的支持，目前的实现也勉强够用（我自己是用来跑PaddleDetection的训练，没遇到过什么错误）
如果要完全解决上述问题，应该要为那些native类型补充缺失的pickle相关接口，这个工程量就相对较大了

Comment (luotao1) [reply python/paddle/fluid/reader.py:493]: @wwqgtxx 
> 如果要完全解决上述问题，应该要为那些native类型补充缺失的pickle相关接口，这个工程量就相对较大了

- 确实这个工程量相对比较大了，如果有兴趣的话，可以帮我们来修复，非常欢迎。
- 我们正在运营一个[Paddle Framework Contributor Club (PFCC)](https://github.com/PaddlePaddle/community/tree/master/pfcc) 组织，会通过定期分享技术知识与发布开发者主导任务的形式持续为飞桨框架做贡献，详情可见 https://github.com/luotao1 主页说明。  
- 如何加入：可以提交 PR 至 [Paddle](https://github.com/PaddlePaddle/Paddle)，代码合入后，我们会邀请你加入 PFCC。非常期待你的加入～

来自飞桨PFCC

Comment (ZHUI) [reply python/paddle/fluid/reader.py:493]: @luotao1 @wwqgtxx 这个PR还有希望继续更新吗？目前pickle已经初步支持，需要补充 mac 和win上的一些传输处理。这块我可以提供一定指导。

Comment (ZHUI) [reply python/paddle/fluid/reader.py:493]: https://github.com/PaddlePaddle/Paddle/pull/47025/files  https://github.com/PaddlePaddle/Paddle/pull/48179/files

这个pr将 multiprocessing 修改为默认引入。

Comment (ZHUI) [reply python/paddle/fluid/reader.py:493]: https://github.com/PaddlePaddle/Paddle/pull/37302 这个pr是支持 multiprocessing 支持的pr，可以根据这个修改一些支持一下 mac or win 平台。

Comment (luotao1) [reply python/paddle/fluid/reader.py:493]: > 这块我可以提供一定指导。

@ZHUI 太好啦

> 这个PR还有希望继续更新吗？

@wwqgtxx 请问你还有时间和兴趣继续更新么，如果可以的话，稍后我可以拉个群

**Pull request closed without merging**